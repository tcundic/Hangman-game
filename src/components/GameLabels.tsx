import React from "react";
import Props from "../models/Props";
import Game from "../models/game";
import Highscore from "../models/highscore";
import {connect} from "react-redux";

export const GameLabels = (props: Props) => {
    return (
        <div>
            <div className="is-size-3 has-text-weight-bold">Player: {props.game?.userName}</div>
            <div className="is-size-3 has-text-weight-bold">Errors: {props.game?.errors}</div>
        </div>
    );
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game
    };
};

export default connect(mapStateToProps)(GameLabels);
