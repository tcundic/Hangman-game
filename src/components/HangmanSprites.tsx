import React from "react";
import {connect} from "react-redux";
import Props from "../models/Props";
import Game from "../models/game";
import Highscore from "../models/highscore";
import Sprite from "../utils/Sprite";

export class HangmanSprites extends React.Component<Props> {

    getSpritePosition = (numberOfErrors: number | undefined) => {
        switch(numberOfErrors) {
            case 0:
                return {x: 0, y: 0, width: 120, height: 170};
            case 1:
                return {x: 120, y: 0, width: 130, height: 170};
            case 2:
                return {x: 250, y: 0, width: 130, height: 170};
            case 3:
                return {x: 370, y: 0, width: 130, height: 170};
            case 4:
                return {x: 0, y: 170, width: 120, height: 170};
            case 5:
                return {x: 120, y: 170, width: 125, height: 170};
            case 6:
                return {x: 240, y: 170, width: 130, height: 170};
            case 7:
                return {x: 370, y: 170, width: 130, height: 170};
            default:
                return {x: 370, y: 170, width: 130, height: 170};
        }
    };

    render() {
        const {x, y, width, height} = this.getSpritePosition(this.props.game?.errors);
        return <Sprite filename="./hangman.png" x={x} y={y} width={width} height={height} />
    }
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game
    };
};

export default connect(mapStateToProps)(HangmanSprites);