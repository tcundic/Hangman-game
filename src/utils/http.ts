import axios from 'axios';
import Highscore from "../models/highscore";

const QUOTABLE_API = 'https://api.quotable.io';
const INGEMARK_API = 'https://my-json-server.typicode.com/stanko-ingemark/hang_the_wise_man_frontend_task'

const httpClient = (url: string) => axios.create({
    baseURL: url,
    headers: {
        "Content-type": "application/json"
    }
});

const getRandomQuote = async () => {
    try {
        const res = await httpClient(QUOTABLE_API).get<{ _id: string, content: string }>("/random");
        console.log(res);
        return {id: res.data._id, content: res.data.content};
    } catch (e) {
        console.error("Error while fetching quote!");
    }
};

const sendStatistics = (
    {quoteId, length, uniqueCharacters, userName, errors, duration}:
        { quoteId: string, length: number, uniqueCharacters: number, userName: string, errors: number, duration: number }) => {
    try {
        httpClient(INGEMARK_API).post("/highscores", {
            quoteId,
            length,
            uniqueCharacters,
            userName,
            errors,
            duration
        });
    } catch (e) {
        console.error("Error while sending score!");
    }
};

const getHighscores = async () => {
    try {
        const res = await httpClient(INGEMARK_API).get<Array<Highscore>>("/highscores");
        return res;
    } catch (e) {
        console.error("Error while fetching highscores!");
    }
};

export {getRandomQuote, sendStatistics, getHighscores};