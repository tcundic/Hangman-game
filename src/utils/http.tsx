import axios from 'axios';

const QUOTABLE_API = 'https://api.quotable.io';
const INGEMARK_API = 'https://my-json-server.typicode.com/stanko-ingemark/hang_the_wise_man_frontend_task'

const httpClient = (url: string) => axios.create({
    baseURL: url,
    headers: {
        "Content-type": "application/json"
    }
});

const getRandomQuote = async () => {
    try {
        const res = await httpClient(QUOTABLE_API).get<{_id: string, content: string}>("/random");
        return {id: res.data._id, content: res.data.content};
    } catch (e) {
        console.error("Error while fetching quote!");
    }
};

const sendStatistics = (
    {quoteId, length, uniqueCharacters, userName, errors, duration}:
        {quoteId: string, length: number, uniqueCharacters: number, userName: string, errors: number, duration: number}) => {
    httpClient(INGEMARK_API).post("/highscores", {
        quoteId,
        length,
        uniqueCharacters,
        userName,
        errors,
        duration
    });
};

export {getRandomQuote, sendStatistics};