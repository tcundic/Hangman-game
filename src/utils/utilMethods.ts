import {getRandomQuote} from "./http";
import Game from "../models/game";
import Highscore from "../models/highscore";
import {resetCurrentGame, setQuote} from "../actions/games";

const getUniqueLetters = (quote: string) => new Set(quote.replace(/[^A-Za-z]/g, "").toLowerCase());

const getNewQuote = async () => {
    const res = await getRandomQuote();
    return res;
}

const isFinishedGame = (quote: string, revealedLetters: Array<string>) => {
    const uniqueCharacters = getUniqueLetters(quote);
    return uniqueCharacters.size === revealedLetters.length;
};

const mapDispatchToProps = (dispatch: Function) => {
    return {
        startGame: (quoteId: string, content: string) => {
            dispatch(resetCurrentGame());
            dispatch(setQuote(quoteId, content));
        },
        dispatch
    }
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game,
        highscores: state.highscores
    };
};

export {getUniqueLetters, getNewQuote, isFinishedGame, mapStateToProps, mapDispatchToProps};
