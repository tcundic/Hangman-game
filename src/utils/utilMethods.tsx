import {getRandomQuote} from "./http";
import Game from "../models/game";
import Highscore from "../models/highscore";
import {resetCurrentGame, setQuote, startTiming} from "../actions/games";

const getUniqueLetters = (quote: string) => new Set(quote.replace(/[^A-Za-z]/g, "").toLowerCase());

const getNewQuote = async () => {
    const res = await getRandomQuote();
    if (res) {
        const {id, content} = res;
        return {id, content};
    }
}

const isFinishedGame = (quote: string, revealedLetters: Array<string>) => {
    const uniqueCharacters = getUniqueLetters(quote);
    return uniqueCharacters.size === revealedLetters.length;
};

const mapDispatchToProps = (dispatch: Function) => {
    return {
        resetGame: (quoteId: string, content: string) => {
            dispatch(resetCurrentGame());
            dispatch(setQuote(quoteId, content));
        },
        startGame: (quoteId: string, content: string) => {
            dispatch(setQuote(quoteId, content));
            dispatch(startTiming());
        },
        dispatch
    }
};

const mapStateToProps = (state: { game: Game, highscores: Array<Highscore> }) => {
    return {
        game: state.game,
        highscores: state.highscores
    };
};

export {getUniqueLetters, getNewQuote, isFinishedGame, mapStateToProps, mapDispatchToProps};
