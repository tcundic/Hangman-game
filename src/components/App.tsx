import React from 'react';
import {connect} from 'react-redux';
import {Navigate} from "react-router-dom";
import Props from "../models/Props";
import {setUserName} from "../actions/games";
import {mapDispatchToProps, mapStateToProps} from "../utils/utilMethods";

export class App extends React.Component<Props> {
    state = {
        username: ''
    };

    onUsernameSave = (): void => {
        this.props.dispatch(setUserName(this.state.username));
    };

    onUsernameChange = (e: React.ChangeEvent<any>): void => {
        const username = e.target.value;
        this.setState(() => ({username}));
    }

    render() {
        if (this.props.game?.userName) {
            return <Navigate to="/play" />;
        }

        return (
            <div className="container mt-6">
                <label className="label is-large">Username</label>
                <div className="field is-grouped">
                    <p className="control">
                        <input
                            className="input is-large"
                            type="text"
                            placeholder="Enter username"
                            value={this.state.username}
                            onChange={this.onUsernameChange}
                        />
                    </p>
                    <p className="control">
                        <button className="button is-info is-large" onClick={this.onUsernameSave}>
                            Save username
                        </button>
                    </p>
                </div>
            </div>
        );
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(App);