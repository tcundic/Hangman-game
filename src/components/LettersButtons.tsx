import React from "react";
import {connect} from "react-redux";
import Props from "../models/Props";
import LetterButton from "./LetterButton";
import {stopTiming, useLetter} from "../actions/games";
import {mapDispatchToProps, mapStateToProps, isFinishedGame} from "../utils/utilMethods";
import {sendStatistics} from "../utils/http";

export class LettersButtons extends React.Component<Props> {
    letters: Array<string>;

    constructor(props: Props) {
        super(props);

        const alpha: Array<number> = Array.from(Array(26)).map((e, i) => i + 65);
        this.letters = alpha.map((x) => String.fromCharCode(x));
    }

    componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<{}>, snapshot?: any) {
        if (isFinishedGame(this.props.game?.content || "", this.props.game?.revealedLetters || [])) {

            const quoteId = this.props.game?.quoteId || "";
            const length = this.props.game?.length || 0;
            const uniqueCharacters = this.props.game?.uniqueCharacters || 0;
            const userName = this.props.game?.userName || "";
            const errors = this.props.game?.errors || 0;
            const duration = this.props.game?.duration || 0;

            sendStatistics({
                quoteId,
                length,
                uniqueCharacters,
                userName,
                errors,
                duration
            });
        }
    }

    onLetterClick = (e: React.ChangeEvent<any>) => {
        this.props.dispatch(useLetter(e.target.textContent));

        if (isFinishedGame(this.props.game?.content || "", this.props.game?.revealedLetters || [])) {
            this.props.dispatch(stopTiming());
        }
    };

    render() {
        return (
            <div className="columns mx-4 my-2">
                {this.letters.map(letter =>
                    <LetterButton onClick={this.onLetterClick} key={letter} letter={letter} used={!!this.props.game?.usedLetters.includes(letter)} />
                )}
            </div>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(LettersButtons);