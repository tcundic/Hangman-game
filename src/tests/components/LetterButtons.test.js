import React from "react";
import { shallow } from "enzyme";
import { LettersButtons } from "../../components/LettersButtons";
import * as http from "../../utils/http";
import * as utils from "../../utils/utilMethods";

const dispatch = jest.fn();
const startGame = () => {
};

const game = {
    quoteId: "",
    content: "",
    userName: "",
    duration: 0,
    errors: 0,
    length: 0,
    uniqueCharacters: 0,
    usedLetters: [],
    revealedLetters: []
};

const highscores = [];

test('should render LetterButtons correctly', () => {
    const wrapper = shallow(<LettersButtons dispatch={dispatch} startGame={startGame} game={game}
                                            highscores={highscores}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should send statistics on game finish', () => {
    const wrapper = shallow(<LettersButtons dispatch={dispatch} startGame={startGame} game={game}
                                            highscores={highscores}/>);

    const sendStatistics = jest.spyOn(http, 'sendStatistics');
    sendStatistics.mockReturnValue('');

    wrapper.setProps({
        game: {
            quoteId: "123",
            content: "New quote",
            userName: "username",
            duration: 100,
            errors: 2,
            length: 9,
            uniqueCharacters: 7,
            usedLetters: ['n', 'x', 'y', 'e', 'w', 'q', 'u', 'o', 't'],
            revealedLetters: ['n', 'e', 'w', 'q', 'u', 'o', 't']
        }
    });

    expect(sendStatistics).toHaveBeenCalled();
});

test('should dispatch events on click', () => {
    const sendStatistics = jest.spyOn(utils, 'isFinishedGame');
    sendStatistics.mockReturnValue(true);

    const mockDate = new Date(1643919358043);
    const spy = jest
        .spyOn(global, 'Date')
        .mockImplementation(() => mockDate);

    const wrapper = shallow(<LettersButtons
        dispatch={dispatch}
        startGame={startGame}
        game={game}
        highscores={highscores}/>);

    wrapper.find('LetterButton').first().simulate('click', {
        target: {
            textContent: 'A'
        }
    });

    expect(dispatch).toHaveBeenCalledWith({
        type: 'USE_LETTER',
        usedLetter: 'A'
    });

    expect(dispatch).toHaveBeenCalledWith({
        type: 'STOP_TIMING',
        endTime: 1643919358043
    });
});
