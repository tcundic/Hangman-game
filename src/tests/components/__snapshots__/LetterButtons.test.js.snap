// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LetterButtons correctly 1`] = `
<div
  className="columns mx-4 my-2"
>
  <LetterButton
    key="A"
    letter="A"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="B"
    letter="B"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="C"
    letter="C"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="D"
    letter="D"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="E"
    letter="E"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="F"
    letter="F"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="G"
    letter="G"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="H"
    letter="H"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="I"
    letter="I"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="J"
    letter="J"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="K"
    letter="K"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="L"
    letter="L"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="M"
    letter="M"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="N"
    letter="N"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="O"
    letter="O"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="P"
    letter="P"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="Q"
    letter="Q"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="R"
    letter="R"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="S"
    letter="S"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="T"
    letter="T"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="U"
    letter="U"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="V"
    letter="V"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="W"
    letter="W"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="X"
    letter="X"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="Y"
    letter="Y"
    onClick={[Function]}
    used={false}
  />
  <LetterButton
    key="Z"
    letter="Z"
    onClick={[Function]}
    used={false}
  />
</div>
`;
